klippy/kinematics/extruder.py

```diff
@@ -185,6 +191,29 @@ def cmd_SET_PRESSURE_ADVANCE(self, params):
                    pressure_advance, smooth_time))
         self.printer.set_rollover_info(self.name, "%s: %s" % (self.name, msg))
         gcode.respond_info(msg, log=False)
+    cmd_GET_E_STEP_DISTANCE_help = "Get extruder step distance"
+    def cmd_GET_E_STEP_DISTANCE(self, params):
+        gcode = self.printer.lookup_object('gcode')
+        step_dist = self.stepper.get_step_dist()
+        gcode.respond_info("%s E step distance: %f" % (self.name, step_dist))
+    cmd_SET_E_STEP_DISTANCE_help = "Set extruder step distance"
+    def cmd_SET_E_STEP_DISTANCE(self, params):
+        gcode = self.printer.lookup_object('gcode')
+        toolhead = self.printer.lookup_object('toolhead')
+        dist = gcode.get_float('DISTANCE', params, 0.)
+        old_step_dist = self.stepper.get_step_dist()
+        if abs((dist - old_step_dist)/old_step_dist) > 0.25:
+            gcode.respond_info("Failed: step change is outside of"
+                               " safe margin. %f -> %f" %
+                                            (old_step_dist, dist))
+        else:
+            self.stepper.set_step_dist(dist)
+            step_dist = self.stepper.get_step_dist()
+            gcode.respond_info("%s E step distance set: %f" %
+                                                (self.name, step_dist))
+            self.stepper.set_max_jerk(9999999.9, 9999999.9)
+            self.stepper.set_stepper_kinematics(self.sk_extruder)
+            self.stepper.set_trapq(self.trapq)
```

Comment:
These three lines should be removed.

The set_max_jerk() call configures the micro-controller "min_stop_interval" configuration.  That configuration can not be changed at run-time, so it's confusing to alter the setting at run-time.  In this particular case, a 9999999.9 disables the min_stop_interval check, so it's redundant to call again anyway.

I understand the set_stepper_kinematics() call is to populate the step_distance to the "itersolve" code.  That's fine, but it should be done from stepper:set_step_dist().

The set_trapq() call associates the stepper with a "motion queue".  That motion queue isn't changing on a step distance update, so there is no reason to call it here.

Separately, I believe it will be necessary to call `toolhead.flush_step_generation()` prior to updating the step distance.

Reply:
ahh, okay, on the max_jerk I didn't see that check, so makes sense.  I will move the set_stepper_kinematics() as suggested.  Thank you for note about toolhead.flush_step_generation().  Thank makes a lot of sense. I will make those changes and test to make sure it performs as expected.